### PR TITLE
Document that debug request/reply and events follow the DAP specification and document added messages

### PR DESCRIPTION
**Pull request by SylvainCorlay**
Cf JEP on the Jupyter Debugger Protocol. https://github.com/jupyter/enhancement-proposals/pull/47

### CODE REVIEW

File: docs/messaging.rst [hunk 1]
```diff
@@ -1092,11 +1092,61 @@ Message type: ``debug_reply``::
...
+      {
+          'type'  : 'response',
+          'success': bool,
+          'body': {
+              'isStarted' : bool  # whether the debugger is started,
+              'hashMethod' : str  # the hash method for code cell. Default is 'Murmur2',
+              'hashSeed' : str  # the seed for the hashing of code cells,
+              'tmpFilePrefix': str,  # prefix for temporary file names
+              'tmpFileSuffix': str,  # suffix for temporary file names
+              'breakpoints': dict(list(breakpoints)),  # current breakpoints currently registered in the debugger.
```
Comment (blink1073) [inline docs/messaging.rst:1146]: Is `breakpoint` a defined object type from DAP?  If so, can we call it out here?

Comment (SylvainCorlay) [reply docs/messaging.rst:1146]: Yep, this is not correct yet. Iterating on this.

File: docs/messaging.rst [hunk 2]
```diff
@@ -1092,11 +1092,61 @@ Message type: ``debug_reply``::
...
-Debug requests and replies are sent over the `control` channel to prevent queuing
-behind execution requests.
+Debug requests and replies are sent over the `control` channel to prevent queuing behind execution requests.
+
+Additions to the DAP
+~~~~~~~~~~~~~~~~~~~~
+
+The Jupyter debugger protocol makes two additions to the DAP, the `dumpCell` request and response, and the `debugInfo` request and response messages.
+
+In order to support the debugging of notebook cells and of Jupyter consoles, which are not based on source files, we need a messages to submit code to the debugger to which breakpoints can be added.
```
Comment (davidbrochart) [inline docs/messaging.rst:1104]: typo: "a messages"

File: docs/messaging.rst [hunk 3]
```diff
@@ -1092,11 +1092,61 @@ Message type: ``debug_reply``::
...
+
+     {
+          'type' : 'response',
+          'success': bool,
+          'body': {
+              'sourcePath': str  # filename for the dumped source
+          }
+     }
+
+In order to support page reloading, or a client connecting at a later stage, Jupyter kernels must store the state of the debugger (such breakpoints, whether the debugger is currently stopped). The `debugInfo` request is a DAP `Request` with no extra argument.
```
Comment (davidbrochart) [inline docs/messaging.rst:1126]: "such **as** breakpoints"?

File: docs/messaging.rst [hunk 4]
```diff
@@ -1092,11 +1092,61 @@ Message type: ``debug_reply``::
...
+          'command' : 'debugInfo'
+      }
+
+  Content of `debugInfo` response::
+
+      {
+          'type'  : 'response',
+          'success': bool,
+          'body': {
+              'isStarted' : bool  # whether the debugger is started,
```
Comment (davidbrochart) [inline docs/messaging.rst:1141]: comma before comment?

File: docs/messaging.rst [hunk 5]
```diff
@@ -1092,11 +1092,73 @@ Message type: ``debug_reply``::
...
+              'hashMethod' : str,  # the hash method for code cell. Default is 'Murmur2',
+              'hashSeed' : str,  # the seed for the hashing of code cells,
+              'tmpFilePrefix': str,  # prefix for temporary file names
+              'tmpFileSuffix': str,  # suffix for temporary file names
+              'breakpoints': [  # current breakpoints currently registered in the debugger.
+                  {
+                      'source' : str,  # source file
+                      'breakpoints':  list  # list of breakpoints for this source file
+                  }
+              ]
```
Comment (davidbrochart) [inline docs/messaging.rst:1151]: Needs a trailing comma.

File: docs/messaging.rst (hunk 1 above)
Comment (SylvainCorlay) [reply docs/messaging.rst]: I have improved it a bit. Still need to detail the `breakpoint` schema.

**Comment by SylvainCorlay** [comment]
@jtpio @afshin @JohanMabille we went with the list of DAP breakpoint in the debugInfo as discussed in the debugger meeting.

File: docs/messaging.rst [hunk 6]
```diff
@@ -1092,11 +1092,68 @@ Message type: ``debug_reply``::
...
+      {
+          'type' : 'response',
+          'success' : bool,
+          'body' : {
+              'isStarted' : bool,  # whether the debugger is started,
+              'hashMethod' : str,  # the hash method for code cell. Default is 'Murmur2',
+              'hashSeed' : str,  # the seed for the hashing of code cells,
+              'tmpFilePrefix' : str,  # prefix for temporary file names
+              'tmpFileSuffix' : str,  # suffix for temporary file names
+              'breakpoints' : [  # current breakpoints currently registered in the debugger.
```
Comment (jtpio) [inline docs/messaging.rst:1146]:
```suggestion
              'breakpoints' : [  # breakpoints currently registered in the debugger.
```

**Comment by jtpio** [comment]
Looks good :+1:

**Comment by SylvainCorlay** [comment]
Now that https://github.com/jupyter/enhancement-proposals/pull/47 is in, I think we can go ahead with this.

File: docs/messaging.rst [hunk 7]
```diff
@@ -1092,11 +1092,68 @@ Message type: ``debug_reply``::
 
     content = {}
 
-The ``content`` dict can be any JSON information used by debugging frontends
-and kernels.
+The ``content`` dicts of the `debug_request` and `debug_reply` messages respectively follow the specification of the `Request` and `Response` messages from the [Debug Adapter Protocol (DAP)](https://microsoft.github.io/debug-adapter-protocol/implementors/adapters/) as of version 1.39 or later.
```
Comment (Carreau) [inline docs/messaging.rst:1095]: LInks are in MD form, but this is an RST file. Maybe other places to be fixed in the doc ?

File: docs/messaging.rst [hunk 8]
```diff
@@ -1382,7 +1439,7 @@ Message type: ``debug_event``::
 
     content = {}
 
-The ``content`` dict can be any JSON information used by debugging frontends.
+The ``content`` dict follows the specification of the `Event` message from the [Debug Adapter Protocol (DAP)](https://microsoft.github.io/debug-adapter-protocol/implementors/adapters/).
```
Comment (Carreau) [inline docs/messaging.rst:1442]: Same here.

**Comment by SylvainCorlay** [comment]
Fixed the requested change by @Carreau.

**Pull request merged**